### PR TITLE
FEATURE: implement positional array sorter and use it to sort the registries

### PR DESCRIPTION
**Pull request by dimaip**
Resolves: #1044 

Changes to the registry API:
- rename `add` to `set`.
- third parameter to `set` takes a position, implementing full positional array sorter spec
- `getChildren` and `getChildrenAsObject` return all keys starting with a certain string

Example:
```js
containerRegistry.set('LeftSideBar/Top/PageTreeToolbar', PageTreeToolbar);
containerRegistry.set('LeftSideBar/Top/PageTreeSearchbar', PageTreeSearchbar);
containerRegistry.set('LeftSideBar/Top/PageTree', PageTree);
```
Then somewhere from the user code:
```js
containerRegistry.set('LeftSideBar/Top/MyCustomBar', MyCustomBar, 'before LeftSideBar/Top/PageTree');
```
The you can render all containers, sorted, like this:
```js
containerRegistry.getChildren('LeftSideBar/Top').map((Item, key) => <Item key={key}/>);
```

### Differences so far from the orginal PositionalArraySorter:
1. Invalid positions are sorted as if they had no positions and only console warning is displayed. PHP sorter would throw. I think it's better to make frontend code more resilient to errors.
2. Chains of circular dependencies are detected and moved to the end. PHP sorter would hang.
3. `start 12 34` was considered an invalid key. Now it's interpreted as `start 12`
4. PHP sorter is recursive, this one is flat. One can always recursively apply if needed.
5. It was possible to add weights to `before` and `after`, now it's not.

TODO:

- [x] use `getChildren` in feedback handlers
- [x] write tests for the registry

### CODE REVIEW

File: packages/neos-ui-extensibility/src/registry/SynchronousRegistry.js [hunk 1]
```diff
@@ -1,38 +1,74 @@
...
+            result[item.key] = item.value;
+        });
+        return result;
+    }
+
+    getChildren(searchKey) {
+        return this._getChildrenWrapped(searchKey).map(item => item.value);
     }
 
     has(key) {
```
Comment (skurfuerst) [inline packages/neos-ui-extensibility/src/registry/SynchronousRegistry.js]: can you add docs to all the public methods here, how they shall be used?

File: packages/neos-ui-extensibility/src/registry/SynchronousRegistry.js [hunk 2]
```diff
@@ -1,38 +1,74 @@
...
-        this._keys = [];
+        this._registry = [];
     }
 
-    add(key, value) {
-        // TODO: implement position in registry!
-        // TODO: "key" must be string!
-        this._registry[key] = value;
-        this._keys.push(key);
+    set(key, value, position = 0) {
```
Comment (skurfuerst) [inline packages/neos-ui-extensibility/src/registry/SynchronousRegistry.js]: I am not sure we should allow having the same key twice. I think this should always be an error, as it makes the remaining API inconsistent / indeterminate

File: packages/utils-helpers/src/positionalArraySorter.spec.js [hunk 3]
```diff
@@ -0,0 +1,179 @@
+import positionalArraySorter from './positionalArraySorter';
+
+test(`Middle`, () => {
```
Comment (skurfuerst) [inline packages/utils-helpers/src/positionalArraySorter.spec.js]: I suppose you ported the tests from the PHP side? (Just asking :-) )

File: packages/neos-ui-extensibility/src/registry/SynchronousRegistry.js (hunk 2 above)
Comment (dimaip) [reply packages/neos-ui-extensibility/src/registry/SynchronousRegistry.js]: Great point!

File: packages/neos-ui-extensibility/src/registry/SynchronousRegistry.js (hunk 1 above)
Comment (dimaip) [reply packages/neos-ui-extensibility/src/registry/SynchronousRegistry.js]: Will do.

File: packages/utils-helpers/src/positionalArraySorter.spec.js (hunk 3 above)
Comment (dimaip) [reply packages/utils-helpers/src/positionalArraySorter.spec.js]: Erm... honestly... no. Will check how they compare.

File: packages/neos-ui-extensibility/src/registry/SynchronousRegistry.js (hunk 2 above)
Comment (dimaip) [reply packages/neos-ui-extensibility/src/registry/SynchronousRegistry.js]: On the other hand, what about if the user wants to override something in our UI? E.g. replace the tree with smth else?

**Comment by skurfuerst** [comment]
perfect :) Awesome job!

**Pull request merged**